### PR TITLE
New DefinePlugin options

### LINKED ISSUE #4724: Document webpack change: DefinePlugin takes care of module invalidation when values have changed
*A pull request by @sokra was merged and maintainers requested a documentation change.*

See pull request: https://github.com/webpack/webpack/pull/12896

---

The options of the DefinePlugin no longer have to be covered by build dependencies

**What kind of change does this PR introduce?**
feature

**Did you add tests for your changes?**
yes

**Does this PR introduce a breaking change?**
no

**What needs to be documented once your changes are merged?**
* The constant options of the DefinePlugin do not have to be covered by build dependencies.
  * but for the `DefinePlugin.runtimeValue` generator function build dependencies need to be used
* `DefinePlugin.runtimeValue` has now an options argument as second argument
  * `options.fileDependencies: string[]` list of files the function depends on
  * `options.contextDependencies: string[]` list of directories the function depends on
  * `options.missingDependencies: string[]` list of not existing files the function depends on
  * `options.buildDependencies: string[]` list of build dependencies the function depends on
  * `options.version: string | () => string` version of the function

### PR DESCRIPTION
**Pull request by anshumanv**
Fix #4724

### CODE REVIEW

**Review by montogeek — DISMISSED** [review]

File: src/content/plugins/define-plugin.md [hunk 1]
```diff
@@ -104,13 +104,22 @@ It is possible to define variables with values that rely on files and will be re
...
 - The first argument of the `webpack.DefinePlugin.runtimeValue` is a `function` that should return the value to be assigned to the definition.
-- The second argument is an array of file paths to watch for. Pass `true` instead of `[string]` here to flag the module as uncacheable.
+- The second argument is an options object argument with the following properties
+  - `options.fileDependencies: string[]` list of files the function depends on
+  - `options.contextDependencies: string[]` list of directories the function depends on
+  - `options.missingDependencies: string[]` list of not existing files the function depends on
+  - `options.buildDependencies: string[]` list of build dependencies the function depends on
+  - `options.version: string | () => string` version of the function
+
+W> The constant options of the DefinePlugin do not have to be covered by build dependencies but for the `DefinePlugin.runtimeValue` generator function build dependencies need to be used.
```
Comment (chenxsan) [inline src/content/plugins/define-plugin.md:114]:
```suggestion
```
Let's remove this as it's confusing without enough context.

File: src/content/plugins/define-plugin.md [hunk 2]
```diff
@@ -104,13 +104,22 @@ It is possible to define variables with values that rely on files and will be re
 Arguments:
 
 - The first argument of the `webpack.DefinePlugin.runtimeValue` is a `function` that should return the value to be assigned to the definition.
```
Comment (chenxsan) [inline src/content/plugins/define-plugin.md:106]: Let's add the signature https://github.com/webpack/webpack/blob/master/types.d.ts#L2408 for the first argument as well.

**Comment by anshumanv** [comment]
Done all

**Comment by chenxsan** [comment]
image

**Comment by chenxsan** [comment]
While I appreciate your help documenting those new features, please do not just copy&paste from the original issue, most of the time it doesn't really make sense. Take this pull request for example, the time I spent reviewing is much more than the time I file another pull request.

**Comment by anshumanv** [comment]
Thanks for help

**Review by chenxsan — APPROVED** [review]

**Pull request merged**